docs/mode-reference.rst

```diff
@@ -154,16 +154,40 @@ for one thing like string in single or double quotes.
 begin
 ^^^^^
 
-- **type**: regexp
+- **type**: regexp or array of regexp
 
 Regular expression starting a mode. For example a single quote for strings or two forward slashes for C-style comments.
 If absent, ``begin`` defaults to a regexp that matches anything, so the mode starts immediately.
 
 
+You can also pass an array when you need to individually highlight portions of the match with different classes:
+
+::
+
+  {
+  begin: [
+    /function!/,
+    /\s+/,
+    hljs.IDENT_RE
+  ],
+  className: {
+    1: "keyword",
+    3: "title"
+  },
+  }
+
+This would highlight ``function!`` as a ``keyword`` while highlighting the name
+of the function as ``title``. The space(s) between would be matched, but not
+highlighted.
+
+*Note: Internally this magic happens by wrapping each regex within it's own
+capture group.  If your regexes include any groups they must all be
+non-capturing, ie:* ``(?:regex)``.
```

Comment:
Is there a way to detect if a RegExp object has capture groups inside of it? If not, I feel like we should specify that using capture groups with a traditional RegExp object will __not__ work since we're only checking if `begin` is an array. Add a link to #3095?

Reply:
>  If not, I feel like we should specify that using capture groups with a traditional RegExp object will not work

I thought that's what it already says.  

> Add a link to #3095?

That's a great idea.

Reply:
To me, this says that if I want to organize my regex and have groups, to always use non-capturing groups. Then it tells me that when an array is given, it will internally convert it to capturing groups for me. It doesn't tell me that I shouldn't/can't use capturing groups right now and expect this behavior.

Reply:
I thought about that (the necessity of array), but there are all sorts of reasons to use capture groups than multi-class... you can use capture groups as much as you want, just NOT with multi-match... 

The other problem with letting someone specify their own capture groups is you have to have ALL of them... you can't leave some content outside a group... so the array allows us to FORCE this behavior since we're in charge of the wrapping... and the more I use I think it's also kind of nice for people who like clean syntax more than pure regex.

I could take another look, but if you see the confusion could you just offer up what you think it should say? Since I plan to fix it soon anyways I'm not wanting to spend much time here.

Reply:
Oh wait. I forgot about this one. One sec!